### PR TITLE
Improve Clarity of the `Build forms with API` Recipe

### LINKED ISSUE #5777: Verify Code in Build Forms recipe
### 📚 Subject area/topic

Building forms recipe

### 📋 Page(s) affected (or suggested, for new content)

https://docs.astro.build/en/recipes/build-forms-api/

### 📋 Description of content that is out-of-date or incorrect

We had two people leave feedback regarding this recipe. Would be great if someone would review it, try it fresh and see if it needs updating or correcting!

Comment 1:
Following the code from https://docs.astro.build/en/recipes/build-forms-api/ and having a project with SSR enabled, the code for the submit function that does prevent default cant be reached

Comment 2:
Suggest pointing out the Vue component for the "build-forms-api" page needs "client:*" within the Astro page.

### 🖥️ Reproduction in StackBlitz (if reporting incorrect content or code samples)

_No response_

### PR DESCRIPTION
**Pull request by VoxelMC**
#### Description (required)

- Reordered recipe steps so that you don't jump back and forth between files.
- Added diffs to the code examples to make the changes more obvious (the code is long and changes didn't jump out before)
- Added a step to make sure readers don't forget to use a `client:*` directive to make their form submission work properly.

Affected page: [Modified](https://docs-git-fork-voxelmc-improve-build-forms-api-astrodotbuild.vercel.app/en/recipes/build-forms-api/#_top)

#### Related issues & labels (optional)

- Closes #5777 
- Suggested label: Improve Documentation

### CODE REVIEW

**Comment by VoxelMC** [comment]
Oh, looks like I did some git wackiness, time to fix that...

**Comment by VoxelMC** [comment]
> Oh, looks like I did some git wackiness, time to fix that...

✅ 😅

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 1]
```diff
@@ -17,9 +17,41 @@ This recipe shows you how to send form data to an API endpoint and handle that d
 
 ## Recipe
 
-1. Create a form component using your UI framework. Each input should have a `name` attribute that describes the value of that input.
+1. First, create a `POST` API endpoint at `/api/feedback` that will receive the form data. Use `request.formData()` to process it. Be sure to validate the form values before you use them.
```
Comment (sarah11918) [inline src/content/docs/en/recipes/build-forms-api.mdx:20]:
```suggestion
1. Create a `POST` API endpoint at `/api/feedback` that will receive the form data. Use `request.formData()` to process it. Be sure to validate the form values before you use them. 
```

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 2]
```diff
@@ -133,187 +165,119 @@ This recipe shows you how to send form data to an API endpoint and handle that d
 
     </UIFrameworkTabs>
 
-  2. Create the `POST` API endpoint that will receive the form data. Use `request.formData()` to process it. Be sure to validate the form values before you use them. 
+
+3. Now, create a function that accepts a submit event, then pass it as a `submit` handler to your form. In the function, call `preventDefault` on the event to override the browser's default submission process.
```
Comment (sarah11918) [inline src/content/docs/en/recipes/build-forms-api.mdx:169]:
```suggestion
3. Create a function that accepts a submit event, then pass it as a `submit` handler to your form. 

    In the function:
    - Call `preventDefault()` on the event to override the browser's default submission process.
    - Create a `FormData` object and send it in a `POST` request to your endpoint using `fetch()`.
```

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 3]
```diff
@@ -133,187 +165,119 @@ This recipe shows you how to send form data to an API endpoint and handle that d
...
-        const email = data.get("email");
-        const message = data.get("message");
-        // Validate the data - you'll probably want to do more than this
-        if (!name || !email || !message) {
-          return new Response(
-            JSON.stringify({
-              message: "Missing required fields",
-            }),
-            { status: 400 }
+    Then, create a `FormData` object and send it in a `POST` request to your endpoint using `fetch`.
```
Comment (sarah11918) [inline src/content/docs/en/recipes/build-forms-api.mdx:171]:
```suggestion
```

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 4]
```diff
@@ -133,187 +165,119 @@ This recipe shows you how to send form data to an API endpoint and handle that d
...
-          return new Response(
-            JSON.stringify({
-              message: "Missing required fields",
-            }),
-            { status: 400 }
+    Then, create a `FormData` object and send it in a `POST` request to your endpoint using `fetch`. 
+
+    <UIFrameworkTabs>
+      <Fragment slot="preact">
+        ```tsx title="src/components/FeedbackForm.tsx" "/api/feedback" add={1, 4-17, 34} add="onSubmit={submit}"
```
Comment (sarah11918) [inline src/content/docs/en/recipes/build-forms-api.mdx:175]:
```suggestion
        ```tsx title="src/components/FeedbackForm.tsx" "/api/feedback" add={1, 4-17, 34} add="onSubmit={submit}" "formData" "e.preventDefault();"
```

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 5]
```diff
@@ -327,52 +291,136 @@ This recipe shows you how to send form data to an API endpoint and handle that d
...
+            <button>Send</button>
+            <p v-if="responseMessage">{{ responseMessage }}</p>
+          </form>
+        </template>
+        ```
+      </Fragment>
+
+    </UIFrameworkTabs>
+
+4. Test out your new form component by importing and displaying it on a page! Be sure to use a `client:*` directive to ensure that the form logic is hydrated when you want it to be.
```
Comment (sarah11918) [inline src/content/docs/en/recipes/build-forms-api.mdx:380]:
```suggestion
4. Import and include your `<FeedbackForm />` component on a page. Be sure to use a `client:*` directive to ensure that the form logic is hydrated when you want it to be.
```

**Review by sarah11918 — COMMENTED** [review]
This looks great! Nicely done!  :partying_face: 

See what you think about my suggestions below, and also, I couldn't suggest it here because it was too early in the page, but is the requirement here *still* `output: 'server'` or would 'hybrid' work, too? This recipe was probably written before hybrid existed, so just want to make sure! (And, if hybrid is OK, then we'd need a note that this page should include `export const prerender = false`)

File: src/content/docs/en/recipes/build-forms-api.mdx (hunk 4 above)
Comment (VoxelMC) [reply src/content/docs/en/recipes/build-forms-api.mdx:178]: Shall I do this for all the frameworks?

Comment (sarah11918) [reply src/content/docs/en/recipes/build-forms-api.mdx:178]: If you think it looks good, then yeah!

**Comment by VoxelMC** [comment]
> This looks great! Nicely done!  :partying_face: 
> 
> See what you think about my suggestions below, and also, I couldn't suggest it here because it was too early in the page, but is the requirement here *still* `output: 'server'` or would 'hybrid' work, too? This recipe was probably written before hybrid existed, so just want to make sure! (And, if hybrid is OK, then we'd need a note that this page should include `export const prerender = false`)

I will do a test to ensure this works!

**Comment by VoxelMC** [comment]
> > This looks great! Nicely done!  🥳
> > See what you think about my suggestions below, and also, I couldn't suggest it here because it was too early in the page, but is the requirement here _still_ `output: 'server'` or would 'hybrid' work, too? This recipe was probably written before hybrid existed, so just want to make sure! (And, if hybrid is OK, then we'd need a note that this page should include `export const prerender = false`)
> 
> I will do a test to ensure this works!

Reflected in my last commit, `output: 'hybrid'` does not affect the recipe, and `export prerender =  true|false;` doesn't either!

File: src/content/docs/en/recipes/build-forms-api.mdx [hunk 6]
```diff
@@ -327,52 +294,136 @@ This recipe shows you how to send form data to an API endpoint and handle that d
...
+    </UIFrameworkTabs>
+
+4. Import and include your `<FeedbackForm />` component on a page. Be sure to use a `client:*` directive to ensure that the form logic is hydrated when you want it to be.
+    
+    <UIFrameworkTabs>
+        <Fragment slot="preact">
+        ```astro title="src/pages/index.astro" "client:load"
+        ---
+        import FeedbackForm from "../components/FeedbackForm"
+        ---
```
Comment (VoxelMC) [inline src/content/docs/en/recipes/build-forms-api.mdx:390]:
```suggestion
        ---
        export const prerender = false;
        import FeedbackForm from "../components/FeedbackForm"
        ---
```

Comment (VoxelMC) [reply src/content/docs/en/recipes/build-forms-api.mdx:390]: @sarah11918 if preferred, I can add this to the code snippet at the end as well.

Comment (sarah11918) [reply src/content/docs/en/recipes/build-forms-api.mdx:390]: Ah, I guess the *page itself* doesn't need to be on-demand rendered as long as the API endpoint is. So, in fact, in hybrid mode, shouldn't `feedback.ts` need `export const prerender = false;`?

Comment (sarah11918) [reply src/content/docs/en/recipes/build-forms-api.mdx:390]: SO, if having the statement doesn't actually make a difference, let's not put it in and... see what happens!  :grinning:

**Review by sarah11918 — APPROVED** [review]
It's YOLO Christmas merge, @VoxelMC - my Christmas present to myself. If we need to PR a fix, we need to PR a fix!

Thank you again for being such a great, helpful, active contributor!  :partying_face:

**Comment by VoxelMC** [comment]
> It's YOLO Christmas merge, @VoxelMC - my Christmas present to myself. If we need to PR a fix, we need to PR a fix!
> 
> Thank you again for being such a great, helpful, active contributor!  :partying_face:

Merry Christmas!!! I'll fix it after the holidays - if it needs to be :)

I'm always over the moon to contribute, so I'm glad to be of good service!

**Pull request merged**